orca/producer_test.go

```diff
@@ -0,0 +1,538 @@
+	defer oobLis1.Stop()
+
+	oobLis2 := newTestOOBListener()
+	lisOpts2 := orca.OOBListenerOptions{ReportInterval: reportInterval2}
+
+	oobLis3 := newTestOOBListener()
+	lisOpts3 := orca.OOBListenerOptions{ReportInterval: reportInterval3}
+
+	// Define a load report to send and expect the client to see.
+	loadReportWant := &v3orcapb.OrcaLoadReport{
+		CpuUtilization: 10,
+		MemUtilization: 100,
+		Utilization:    map[string]float64{"bob": 555},
+	}
+
+	// Receive reports and update counts for the three listeners.
+	var reportsMu sync.Mutex
+	var reportsReceived1, reportsReceived2, reportsReceived3 int
+	go func() {
+		for {
+			select {
+			case r := <-oobLis1.loadReportCh:
+				t.Log("Load report 1 received: ", r)
+				if !proto.Equal(r, loadReportWant) {
+					t.Errorf("Unexpected report received: %+v", r)
+				}
+				reportsMu.Lock()
+				reportsReceived1++
+				reportsMu.Unlock()
+			case r := <-oobLis2.loadReportCh:
+				t.Log("Load report 2 received: ", r)
+				if !proto.Equal(r, loadReportWant) {
+					t.Errorf("Unexpected report received: %+v", r)
+				}
+				reportsMu.Lock()
+				reportsReceived2++
+				reportsMu.Unlock()
+			case r := <-oobLis3.loadReportCh:
+				t.Log("Load report 3 received: ", r)
+				if !proto.Equal(r, loadReportWant) {
+					t.Errorf("Unexpected report received: %+v", r)
+				}
+				reportsMu.Lock()
+				reportsReceived3++
+				reportsMu.Unlock()
+			case <-ctx.Done():
+				// Test has ended; exit
+				return
+			}
+		}
+	}()
+
+	// checkReports is a helper function to check the report counts for the three listeners.
+	checkReports := func(r1, r2, r3 int) {
+		t.Helper()
+		for ctx.Err() == nil {
+			reportsMu.Lock()
+			if r1 == reportsReceived1 && r2 == reportsReceived2 && r3 == reportsReceived3 {
+				// Success!
+				reportsMu.Unlock()
+				return
+			}
+			if reportsReceived1 > r1 || reportsReceived2 > r2 || reportsReceived3 > r3 {
+				reportsMu.Unlock()
+				t.Fatalf("received excess reports. got %v %v %v; want %v %v %v", reportsReceived1, reportsReceived2, reportsReceived3, r1, r2, r3)
+				return
+			}
+			reportsMu.Unlock()
+			time.Sleep(10 * time.Millisecond)
+		}
+		t.Fatalf("timed out waiting for reports received. got %v %v %v; want %v %v %v", reportsReceived1, reportsReceived2, reportsReceived3, r1, r2, r3)
+	}
+
+	// Only 1 listener; expect reportInterval1 to be used and expect the report
+	// to be sent and distributed correctly.
+	awaitRequest(reportInterval1)
+	fake.respCh <- loadReportWant
+	checkReports(1, 0, 0)
+
+	// Register listener 2 with a less frequent interval; no need to recreate
+	// stream.  Report should go to both listeners.
+	oobLis2.cleanup = orca.RegisterOOBListener(lao.sc, oobLis2, lisOpts2)
+	fake.respCh <- loadReportWant
+	checkReports(2, 1, 0)
+
+	// Register listener 3 with a more frequent interval; stream is recreated
+	// with the lower interval after the next report is received.  The first
+	// report will go to all three listeners.
```

Comment:
Why is the lower interval the one used? Also what does lower interval mean? Can you reword to say listener 3s more frequent interval...and why... or something along those lines.

Reply:
As above.